server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -304,6 +310,149 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
         return ClusterState.builder(currentState).blocks(blocks).metadata(metadata).routingTable(routingTable.build()).build();
     }
 
+    static ClusterState addIndexBlock(final Index[] indices, final Map<Index, ClusterBlock> blockedIndices,
+                                      final ClusterState currentState, APIBlock block) {
+        final Metadata.Builder metadata = Metadata.builder(currentState.metadata());
+
+        final Set<Index> indicesToAddBlock = new HashSet<>();
+        for (Index index : indices) {
+            metadata.getSafe(index); // to check if index exists
+            if (currentState.blocks().hasIndexBlock(index.getName(), block.block)) {
+                logger.debug("index {} already has block {}, ignoring", index, block.block);
+            } else {
+                indicesToAddBlock.add(index);
+            }
+        }
+
+        if (indicesToAddBlock.isEmpty()) {
+            return currentState;
+        }
+
+        final ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
+        final RoutingTable.Builder routingTable = RoutingTable.builder(currentState.routingTable());
+
+        for (Index index : indicesToAddBlock) {
+            ClusterBlock indexBlock = null;
+            final Set<ClusterBlock> clusterBlocks = currentState.blocks().indices().get(index.getName());
+            if (clusterBlocks != null) {
+                for (ClusterBlock clusterBlock : clusterBlocks) {
+                    if (clusterBlock.id() == block.block.id()) {
+                        // Reuse the existing UUID-based block
+                        indexBlock = clusterBlock;
+                        break;
+                    }
+                }
+            }
+            if (indexBlock == null) {
+                // Create a new UUID-based block
+                indexBlock = createUUIDBasedBlock(block.block);
+            }
+            assert Strings.hasLength(indexBlock.uuid()) : "Block should have a UUID";
+            blocks.addIndexBlock(index.getName(), indexBlock);
+            blockedIndices.put(index, indexBlock);
```

Comment:
I'm really uncomfortable mutating the `blockedIndices` map that was passed in as a parameter. I had to read this over and over to see what was happening. Can we refactor this method so that it becomes part of the return value (maybe using a `Tuple` if we don't want to add an object) so that it's easier to reason about?

Reply:
Good idea, changed  this to a tuple